### PR TITLE
[WIP] [Do not review] [For discussion only atm] Z3 simplification for shape_env guards

### PR DESCRIPTION
**Pull request by voznesenskym**
```
TORCHDYNAMO_DYNAMIC_SHAPES=1  AOT_DYNAMIC_SHAPES=1 python  benchmarks/dynamo/benchmarks.py --backend aot_eager --training  --accuracy --timing --only hf_Reformer
```

### CODE REVIEW

File: torch/_dynamo/guards.py [hunk 1]
```diff
@@ -628,13 +642,25 @@ def direct_equality(a, b):
...
         code = " and ".join(unique(code_parts))
         closure_vars = collections.OrderedDict(
             [
                 ("___guarded_code", self),
                 ("___check_tensors", check_tensors_fn),
                 ("___check_tensors_verbose", check_tensors_verbose_fn),
                 ("tensor_check_names", tensor_check_names),
+                ("Not", not_),
+                ("Or", or_),
+                ("ToReal", to_real_),
```
Comment (ezyang) [inline torch/_dynamo/guards.py:663]: put it in `SYMPY_INTERP`

Comment (voznesenskym) [reply torch/_dynamo/guards.py:663]: I'm going to look at writing a z3 printer to elide this out in favor of just regular py expressions...

**Comment by voznesenskym** [comment]
So, the plan here is to:

1) move this behind a flag
2) Add a ton of tests for various kinds of guards we can encounter
3) Rewrite the calls to not be part of the regular guard extracting contract (should be one, or the other, invoke by dynamo)
4) Rewrite the name remapping to require remapped map (we can do this by making real structures for wrapping the expressions, instead of passing it all alongside it and rediscovering the remap)
5) Fix some of the soundness issues in the algo below (substring match on unsorted by length terms, etc)
6) Sweep and sweep more to figure out where it sits
7) Discuss removal of various strategies this unlocks, upstream sympy simplification removal, accumulating z3, etc etc

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -1041,9 +1043,166 @@ def track_symint(source, val):
...
+                    # assert isinstance(real_value, torch.nn.Module), breakpoint()
+                    new_name = name.replace(".", "_")
+                    remapped_names[name] = new_name
+                    # name = new_name
+                    # breakpoint()
+                    return new_name
+                return name
+
+            if "[" in name:
+                assert "]" in name
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1142]: this 100% doesn't work for something like `[k][n]...`, but have not yet encountered it, still , easy to support and make `[k]` work within it

**Pull request closed without merging**